habitat_baselines/utils/common.py

```diff
@@ -232,70 +229,10 @@ def batch_obs(
     Returns:
         transposed dict of torch.Tensor of observations.
     """
-    batch_t: TensorDict = TensorDict()
-    if cache is None:
-        batch: DefaultDict[str, List] = defaultdict(list)
-
-    obs = observations[0]
-    # Order sensors by size, stack and move the largest first
-    sensor_names = sorted(
-        obs.keys(),
-        key=lambda name: 1
-        if isinstance(obs[name], numbers.Number)
-        else np.prod(obs[name].shape),  # type: ignore
-        reverse=True,
+    return TensorDict.from_tree(observations).map_in_place(
```

Comment:
I need some help understanding what cache is doing. I assume it is for speed, but I wonder if it is still necessary given that we no longer need to process the observations.

Reply:
Still needed. It's primary purpose to cache pinned memory allocations that way the CPU -> pinned memory and pinned memory -> GPU steps needed to upload each observations can be pipelined instead of done sequentially.

Reply:
Ok, I added it back. I think it does what it is supposed to, but I would appreciate if you looked to make sure it uses the cache properly.

Reply:
Seems correct, but I can take a closer look later.


RE your commit message. One nice thing about that cache is that it can't ever decrease performance (expect in very weird cases where the machine doesn't have a few MB of memory to mark as pinned). Calling `.to(device='cuda')` implicitly allocates pinned memory, copies into that memory, and then uploads to the GPU (without `non_blocking=True`). So this is doing the same thing but with `non_blocking=True` to improve perf!

Reply:
I did not know that is what `.to` work. I do know that there is a cost to sending data over and from the GPU to the CPU. I guess the gains are greater when the observations are large images since the cost of sending the data will be higher then. It also explains why the larger observations are processed first.

Reply:
Yeah, that's how CPU -> GPU transfers must be made from pinned memory. That keeps the CPU from paging the memory during the transfer and allows the CPU to be entirely bypassed once the transfer has been initiated. So if the CPU memory isn't already pinned memory, things are first copied into pinned memory and then transferred (really `cudaMemcpy` does this, not `.to`, and it's probably more complicated than just that to support transferring buffers that are larger than the max pinned memory size but nvidia ain't telling).